File: examples/secure/csharp/SqlInjectionFixed.cs
/**
 * ✅ SECURE — Fixed SQL Injection Example (C#)
 *
 * Security Principle : Parameterized Queries (SqlCommand.Parameters)
 * CWE Addressed      : CWE-89  (Improper Neutralization of Special Elements used in an SQL Command)
 * CodeQL query        : cs/sql-injection
 *
 * What Changed:
 *   1. Replaced string concatenation with a parameterized SqlCommand.
 *      Parameters.AddWithValue binds the user input as a typed value,
 *      so the database engine treats it as DATA, never as SQL syntax.
 *   2. Added input validation (null / whitespace check).
 *   3. Replaced raw exception message with a generic 500 response.
 *
 * Compare with: examples/vulnerable/csharp/SqlInjection.cs
 */

using System;
using System.Data.SqlClient;
using Microsoft.AspNetCore.Mvc;

namespace GhasPack.Secure.CSharp
{
    [ApiController]
    [Route("api/[controller]")]
    public class SqlInjectionFixedController : ControllerBase
    {
        private const string ConnectionString =
            "Server=localhost;Database=AppDb;Trusted_Connection=True;";

        // GET /api/sqlinjectionfixed?name=alice
        [HttpGet]
        public IActionResult Search([FromQuery] string name)
        {
            // FIX #1: Input validation — reject missing or blank input
            if (string.IsNullOrWhiteSpace(name))
            {
                return BadRequest("Missing required parameter: name");
            }

            try
            {
                using var connection = new SqlConnection(ConnectionString);
                connection.Open();

                // FIX #2: Parameterized query — the @name placeholder tells
                // ADO.NET to send the query structure and the parameter value
                // to SQL Server SEPARATELY.  The server compiles the query
                // plan once, then binds the parameter as a literal string.
                // Even if 'name' contains ' OR '1'='1, it is treated as a
                // value to compare against, NOT as part of the SQL command.
                // This is WHY parameterized queries prevent SQL injection.
                using var cmd = new SqlCommand(
                    "SELECT * FROM Users WHERE Name = @name", connection);
                cmd.Parameters.AddWithValue("@name", name);

                using var reader = cmd.ExecuteReader();

                var results = new System.Collections.Generic.List<string>();
                while (reader.Read())
                {
                    results.Add($"{reader["Name"]} — {reader["Email"]}");
                }

                return Ok(results);
            }
            catch (SqlException)
            {
                // FIX #3: Generic error — never expose SQL details to callers
                return StatusCode(500, "An internal error occurred. Please try again later.");
            }
        }
    }
}

File: examples/secure/java/SqlInjectionFixed.java
/**
 * ✅ SECURE — Fixed SQL Injection Example
 *
 * Security Principle : Parameterized Queries (Prepared Statements)
 * CWE Addressed      : CWE-89  (Improper Neutralization of Special Elements used in an SQL Command)
 * CodeQL query        : java/sql-injection
 *
 * What Changed:
 *   1. Replaced string concatenation with PreparedStatement and parameter
 *      placeholders (?).  The JDBC driver sends the query structure and
 *      parameter values to the database SEPARATELY, so user input can
 *      never alter the SQL syntax — it is always treated as data.
 *   2. Added input validation (null / empty check) before querying.
 *   3. HTML-encoded output to prevent reflected XSS.
 *   4. Replaced raw error messages with a generic error response.
 *
 * Compare with: examples/vulnerable/java/SqlInjection.java
 */

import java.io.IOException;
import java.io.PrintWriter;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

@WebServlet("/users")
public class SqlInjectionFixed extends HttpServlet {

    private static final String DB_URL = "jdbc:mysql://localhost:3306/appdb";
    private static final String DB_USER = "root";
    private static final String DB_PASS = "password";

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {

        String name = request.getParameter("name");

        // FIX #1: Input validation — reject null or empty input early
        if (name == null || name.trim().isEmpty()) {
            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
            response.getWriter().println("Missing required parameter: name");
            return;
        }

        response.setContentType("text/html");
        PrintWriter out = response.getWriter();

        // FIX #2: Use PreparedStatement with a parameter placeholder (?).
        // The database engine compiles the query structure first, then binds the
        // user-supplied value as a LITERAL string parameter.  Even if the input
        // contains SQL metacharacters like ' OR '1'='1, it is treated as data,
        // NOT as part of the SQL command — this is WHY prepared statements
        // prevent SQL injection.
        String sql = "SELECT * FROM users WHERE name = ?";

        try (Connection connection = DriverManager.getConnection(DB_URL, DB_USER, DB_PASS);
             PreparedStatement ps = connection.prepareStatement(sql)) {

            // FIX #3: Bind the user input as a typed parameter
            ps.setString(1, name);

            try (ResultSet rs = ps.executeQuery()) {
                out.println("<html><body><h2>Search Results</h2><ul>");
                while (rs.next()) {
                    // FIX #4: HTML-encode output to prevent reflected XSS
                    String safeName = htmlEncode(rs.getString("name"));
                    String safeEmail = htmlEncode(rs.getString("email"));
                    out.println("<li>" + safeName + " &mdash; " + safeEmail + "</li>");
                }
                out.println("</ul></body></html>");
            }

        } catch (SQLException e) {
            // FIX #5: Return a generic error message — never leak SQL details
            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
            out.println("<p>An internal error occurred. Please try again later.</p>");
            // Log the real error server-side (not shown to user)
            getServletContext().log("Database error in /users", e);
        }
    }

    /** Minimal HTML entity encoding to prevent XSS in reflected output. */
    private static String htmlEncode(String input) {
        if (input == null) return "";
        return input.replace("&", "&amp;")
                     .replace("<", "&lt;")
                     .replace(">", "&gt;")
                     .replace("\"", "&quot;")
                     .replace("'", "&#x27;");
    }
}
